### DISCUSSION TITLE
Can a function in *luasnip-match* return a snippetNode?

### DISCUSSION THREAD

**Discussion by no-vici** [L3MON4D3/LuaSnip#642] [Q&A] [answered]
Dear LuaSnip lovers,

I'm new to Lua and obviously to LuaSnip.

I would like to have a snippet for ConTEXt Lua that changes the output (which might have an insertNode) based on the text in the first insertNode.

For example:

\start(1) -- (1) is i(1)
i(2, "type here")
\stop(3) -- (3) is rep(1), a copy of i(1)

If i(1) == "itemize", then the output would be:
\startitemize[char] -- char is another insertNode: a, n ...
      \item i(2)
\stopitemize

Bellow is what I could come up with:
```
	s("start", {
		t {"\\start"}, i(1), m(1, 
        function (args)
            local input_text = args[1][1]:lower()
            if input_text == "itemize" then
                -- This function shall return a string, 
                -- returning snippetNodes does not work.
                return "[a] \n\n \\item "
                -- return sn(nil, { t {"["}, i(1, "[a]"), t {"]"}, t {"", "", "\\item "})
            elseif input_text == "placefigure" then
                return "\n\n\t\\externalfigure[fig][opt]"
            elseif input_text == "chapter" 
                or input_text == "section" 
                or input_text == "subsection" then
                return "[title={}]\n\n"
            else
                return '\n\n'
            end
		end), 
        i(2, "type here"), t {"","", "\\stop"}, rep(1),
	}),
```

The function in *luasnip-match* only returns values as strings. I'm stuck at the part that I would like to jump to and edit, e.g, when I type "itemize", "itemize[a]" would come up, then I could jump into the "[a]" and change "a" to "n" or anything else.

With strings as returns of luasnip-match's function, I have to manually jump to "[a]" and change "a" to what I need.

If you have any thought, please kindly share with me.
 
Thank you so much in advance.

**Comment by L3MON4D3** [comment, marked as answer]
Hi!
I think you'll want to use [dynamicNode](https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#dynamicnode), it can generate both text and new insertNodes based on some input.

**Reply by no-vici** [reply]
@L3MON4D3 ,

Thank you for your quick answer. I have tried that but failed badly ;)

```
s("trig", {
        t"text: ", i(1), t{"", "copy: "},
        d(2, 
        function(args)
            if args[1][1] == "itemize" then
                return sn(nil, { t {"["}, i(1, "[a]"), t {"]"}, t {"", "", "\\item "}, i(2, "your item")})
            else
                return '\n\n'
            end
        end, {1}),
    }),
```

When triggered, I got this:

```
Error executing vim.schedule lua callback: ...k/packer/start/LuaSnip/lua/luasnip/nodes/dynamicNode.lua:234: attempt to index lo
cal 'tmp' (a string value)
stack traceback:
        ...k/packer/start/LuaSnip/lua/luasnip/nodes/dynamicNode.lua:234: in function 'update_static'
        .../pack/packer/start/LuaSnip/lua/luasnip/nodes/snippet.lua:819: in function 'update_static'
        .../pack/packer/start/LuaSnip/lua/luasnip/nodes/snippet.lua:744: in function 'fake_expand'
        .../pack/packer/start/LuaSnip/lua/luasnip/nodes/snippet.lua:789: in function 'get_docstring'
        ...e/pack/packer/start/cmp_luasnip/lua/cmp_luasnip/init.lua:35: in function 'get_documentation'
        ...e/pack/packer/start/cmp_luasnip/lua/cmp_luasnip/init.lua:119: in function 'resolve'
       .../pack/packer/start/nvim-cmp/lua/cmp/source.lua:366: in function 'resolve'
        ...e/nvim/site/pack/packer/start/nvim-cmp/lua/cmp/entry.lua:485: in function 'resolve'
        ...re/nvim/site/pack/packer/start/nvim-cmp/lua/cmp/view.lua:234: in function 'fn'
        .../site/pack/packer/start/nvim-cmp/lua/cmp/utils/async.lua:55: in function <.../site/pack/packer/start/nvim-cmp/lua/cm
p/utils/async.lua:53>
```
The problem comes from: args[1][1] == "itemize". I don't quite understand functionNode yet, so it is still very confusing to me.

**Reply by L3MON4D3** [reply]
Ahh, it's just that the function returns a string, `'\n\n'`. Wrapping that in a textNode+snippetNode should do the trick.
(Basically dynamicNode expects its function to always return a `snippetNode`)

**Reply by no-vici** [reply]
@L3MON4D3,

Wow, what can I say? It just works beautifully. Big thanks!

**Reply by L3MON4D3** [reply]
Niiice, you're welcome :)

**Reply by leiserfg** [reply]
You can say "LuaSnip is love"

**Reply by no-vici** [reply]
@leiserfg, 

LuaSnip is a radiant star!

**Reply by no-vici** [reply]
Hope these lines will save ConTEXt users a lot of typing time:

```
s("start", {
		t({ "\\start" }),
		i(1),
		d(2, function(args)
            -- The function inside dynamicNode always return a snippetNode,
            -- returning strings will cause errors.
            -- args[1][1] is the input text from i(1)
			local input_text = args[1][1]
			if input_text == "itemize" then
				return sn(nil, { t({ "[" }), i(1, "a"), t({ "]" }), t({ "", "", "\\item " }), i(2, "your item") })
			elseif input_text == "placefigure" then
				return sn(
					nil,
					{ t({ "", "", "\t\\externalfigure[" }), i(1, "fig"), t({ "][" }), i(2, "opt"), t({ "]" }) }
				)
			elseif input_text == "chapter" or input_text == "section" or input_text == "subsection" then
				return sn(nil, { t({ "[title={" }), i(1, "Title"), t({ "}]" }), t({ "", "", "" }), i(2, "input text") })
			else
				return sn(nil, { t({ "", "" }), i(1, "input text") })
			end
		end, { 1 }),
		t({ "", "", "\\stop" }),
		rep(1),
	}),
```

**Reply by no-vici** [reply]
@L3MON4D3,

Honestly, I have spent the last three days on the problem and could not figure it out. Once again, thank you so so much.

I would love to buy you a cup of coffee ;)

**Reply by L3MON4D3** [reply]
Hah, thank you :D
And feel free to ask earlier! I really don't mind such questions since they're public and will help other users as well :)
Oh, and part of the wiki is meant for examples/usable snippets, definitely add your snippet there for visibility

**Reply by no-vici** [reply]
@L3MON4D3,
You are a very kind person, giving a away all your hard work on LuaSnip and the support along its way to the users for nothing.
I will fine-tune my snippets for ConTEXt and share them with people on LuaSnip's wiki. 

Thank you once more.

**Reply by L3MON4D3** [reply]
Warms my heart to hear this, thank you ❤️❤️
